Rebalance spellbooks

RikerW opened the pull request:
Notable changes:

* delay is now 10*lvl, and the delay slot of the objects.c macro has been removed and mapped to level. delay is still checked in the spell.c code, since it doesn't matter which and in case of a future instant-spell-read intrinisic, or something like that. idk just didn't see a point bothering to remove it, not like it's vanishing from obj->oc_delay anyway.
* Failure rates modified. Base is still 4+int+ulvl/2 - bklvl*2. Modifiers are blessed (+10, down from 100% success), cursed (-10, up from 0% success), ward (+10), lenses (+2). New modifier: current spell memory. If the spell is known by you, you get a modifier that scales linearly up to +10 with the spell memory (so 90% memory is +9).
* Failure effects modified. The format is still rn2(lev), where lev -=1 for being ward. Also decrements effective lev now if the spell is known in general, so a re-read that fails will have nicer effects. The effects are changed to (in order of diff, 0-6 inclusive):
  * 0: teleport you with 3/4 chance, unbless the book / curse the book, if cursed always teles.
  * 1: agg monster.
  * 2: blind for rn1(100, 250) turns (250+rnd(100).
  * 3: book bites you, dealing rn1(5, 3) damage.
  * 4: freezes you for rn1(lvl*2, lvl*2)
  * 5: contact poison, unchanged.
  * 6: explosive energy, messaging changed a tad and now checks 1/2 spell damage because it's funny
* Spellbooks can now be read when you still have memory. Will warn at 10%+ and prompt yn before restoring fully. Infinite spells will always refill fully with no warn prompt.
* Minor change: spellbooks now blank at the end of the reading that sets their charges to 0, instead of the reading after that (which will fail).

RikerW committed:
Rebalance spellbooks

Notable changes:

* delay is now 10*lvl, and the delay slot of the objects.c macro has been removed and mapped to level. delay is still checked in the spell.c code, since it doesn't matter which and in case of a future instant-spell-read intrinisic, or something like that. idk just didn't see a point bothering to remove it, not like it's vanishing from obj->oc_delay anyway.
* Failure rates modified. Base is still 4+int+ulvl/2 - bklvl*2. Modifiers are blessed (+10, down from 100% success), cursed (-10, up from 0% success), ward (+10), lenses (+2). New modifier: current spell memory. If the spell is known by you, you get a modifier that scales linearly up to +10 with the spell memory (so 90% memory is +9).
* Failure effects modified. The format is still rn2(lev), where lev -=1 for being ward. Also decrements effective lev now if the spell is known in general, so a re-read that fails will have nicer effects. The effects are changed to (in order of diff, 0-6 inclusive):
** 0: teleport you with 3/4 chance, unbless the book / curse the book, if cursed always teles.
** 1: agg monster.
** 2: blind for rn1(100, 250) turns (250+rnd(100).
** 3: book bites you, dealing rn1(5, 3) damage.
** 4: freezes you for rn1(lvl*2, lvl*2)
** 5: contact poison, unchanged.
** 6: explosive energy, messaging changed a tad and now checks 1/2 spell damage because it's funny
* Spellbooks can now be read when you still have memory. Will warn at 10%+ and prompt yn before restoring fully. Infinite spells will always refill fully with no warn prompt.
* Minor change: spellbooks now blank at the end of the reading that sets their charges to 0, instead of the reading after that (which will fail).

NeroOneTrueKing commented:
What happens if I write:
Fixes #679 
Fixes #859 
?

Doesn't seem to work :(

RikerW committed:
Can't read the wards on the book of the dead, goes straight to reading.

NeroOneTrueKing commented on src/spell.c:
```diff
@@ -160,72 +158,105 @@ char ilet;
+	if(RoSbook == STUDY_WARD) lev -= 1; //The wardings partially protect you from the magic contained within the book.
 	
-	if(RoSbook == STUDY_WARD) lev -= 1; //The wardings paritally protect you from the magic contained within the book.
+	for (int spell = 0; spell < MAXSPELL; spell++)
```
It lets you declare `int spell` inside the `for()` loop? I thought the compiler didn't like that.

NeroOneTrueKing commented on src/spell.c:
```diff
@@ -160,72 +158,105 @@ char ilet;
-cursed_book(bp)
-	struct obj *bp;
-{
+cursed_book(struct obj *bp){
```
The standard for parameters is to put only their name here, and put their types below:
```
boolean
cursed_book(bp)
struct obj * bp;
{
```

RikerW replied:
K&R style func decs are not used by anybody except for consistency with existing code. I'm not particularly concerned about that, and I don't really feel like keeping it k&r forever is a good idea anyway ([see here, which is my token anti-K&R link](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=48091)).

NeroOneTrueKing commented on src/spell.c:
```diff
@@ -160,72 +158,105 @@ char ilet;
+			}
+			else {
+				if (bp->blessed) {
+					pline_The("book glows brown.");
```
I believe it's technically possible to read a spellbook while hallucinating, so this could instead call hcolor:
`pline_The("book glows %s.", hcolor("brown"));`
And similarly for glowing black.

NeroOneTrueKing commented on src/spell.c:
```diff
@@ -160,72 +158,105 @@ char ilet;
-		    losehp(2*rnd(10)+5, "exploding rune", KILLED_BY_AN);
-		}
+			
+			if (Half_spell_damage) dmg = (dmg+1/2);
```
For extra fun, why not also check `u.uvaul_duration` and reduce damage taken again? 😄

NeroOneTrueKing reviewed (changes requested):
I've just got some nitpicks; this looks good!
You don't have to make changes, they're just things to consider.

RikerW committed:
Add hallucinatory spellbook reading effects

Also, moved those checks to before the failure check, so you don't get get started on reading (I mean come on, if you're "reading the first line over and over" then do you really htink you're even failing to learn the spell?

hallu messages are randomly chosen  with the options being:
* "Wait a minute, where are your reading glasses?" (if you're actually wearing some, goes to the next option)
* "You have absolutely no clue what you just read, but you completely agree with every word."
* "Hm, the herb growing instructions in this book are fascinating."
* "The book contains a detailed recipe for blueberry muffins, complete with the author's signature."

RikerW committed:
deal with gcc warning

RikerW reviewed (commented):
Changes made to address the issues brought up.
